### PR TITLE
Add the capability to call createHash instead of createHmac

### PR DESCRIPTION
**Pull request by dbengsch**
## What does this PR do?

* Add the capability to use createHash instead of createHmac for "old" created hashes
* Stay retro-compatible with older version of the plugin (hmac is used default)
* Allow transparent credentials migration to newer security settings to enhance security
* Add a validation in init method to check if streching and hash encryption are not enabled at the same time
* Introduce a new configuration attribute "encryption" with "hmac" or "hash" as possible values

### How should this be manually tested?

Test has been added to the PR.

### Other changes

* Fixed vulnerabilities as audited by npm

### Boyscout

* README update
* JSDoc

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,8 +12,10 @@ The default configuration is:
 
 ```json
 {
-  "algorithm": "sha256",
-  "digest": "hex"
+  "algorithm": "sha512",
+  "stetching": true,
```
Comment (scottinet) [inline README.md]: can you document what those new properties are doing?

File: lib/passwordManager.js [hunk 2]
```diff
@@ -57,8 +59,10 @@ class PasswordManager {
           resolve(key.toString(this.config.digest));
         });
       }
-      else {
+      else if (!isHash) {
         resolve(crypto.createHmac(algorithm, salt).update(password).digest(this.config.digest));
+      } else {
+        resolve(crypto.createHash(algorithm).update(password).digest(this.config.digest));
```
Comment (scottinet) [inline lib/passwordManager.js]: No salt?

File: README.md (hunk 1 above)
Comment (dbengsch) [reply README.md]: Done

**Comment by scottinet** [comment]
I have an issue with the new `isHash` property. You add it because we only used [createHmac](https://devdocs.io/node~8_lts/crypto#crypto_crypto_createhmac_algorithm_key_options), and now we also need [createHash](https://devdocs.io/node~8_lts/crypto#crypto_crypto_createhash_algorithm_options).
But what if we also need [createCipher](https://devdocs.io/node~8_lts/crypto#crypto_crypto_createcipher_algorithm_password_options), or another `createXxx` method?

I think that the new property should look a bit like this instead:

```js
{
  "algorithm": "sha512",
  "stretching": true,
  "digest": "hex",
  "encryption": "<hmac|hash> (default: hmac)"
}
```

What do you think?

File: lib/passwordManager.js (hunk 2 above)
Comment (dbengsch) [reply lib/passwordManager.js]: Done

**Comment by dbengsch** [comment]
@scottinet good point, I changed it

File: README.md [hunk 3]
```diff
@@ -8,16 +8,23 @@ By default, this plugin is already installed in Kuzzle.
...
+  "stretching": true,
+  "digest": "hex",
+  "encryption": "hmac"
 }
 ```
 
-All the configurations are used to set the behavior of the password hash. `algorithm` must be either `sha256` or `sha512`.
+All the configurations are used to set the behavior of the password hash.
+
+* `algorithm` must be either `sha256` or `sha512`.
```
Comment (scottinet) [inline README.md]: The complete list is available using the following command: `openssl list -digest-algorithms`
There are 40+ algorithms available, the complete list can be retrieved with `crypto.getHashes()`

Proposal change:

```suggestion
* `algorithm`: one of the supported encryption algorithms (run [crypto.getHashes()](https://nodejs.org/dist/latest-v10.x/docs/api/crypto.html#crypto_crypto_gethashes) to get the complete list). Examples: `sha256`, `sha512`, `blake2b512`, `whirlpool`, ...
```

**Comment by scottinet** [comment]
@dbengsch > I took the liberty to change the target branch to be `5-dev`, since I don't think that these changes warrant a hotfix.

I propose that I make a release immediately after that, allowing to update the default plugin version used by Kuzzle in a proper new patch version.

**Comment by dbengsch** [comment]
Yes, sorry, I just about it to @benoitvidis ... I was not sur about it.

**Pull request merged**